src/lib/markbind/src/parser.js

```diff
@@ -29,915 +29,905 @@ const {
-      newBase = combinedBases[bases[0]];
-      const { children } = element;
-      if (children) {
-        const currentBase = calculateNewBaseUrls(element.attribs[ATTRIB_CWF], this.rootPath, this.baseUrlMap);
-        if (currentBase) {
-          if (currentBase.relative !== newBase) {
-            cheerio.prototype.options.xmlMode = false;
-            const newBaseUrl = `{{hostBaseUrl}}/${newBase}`;
-            const rendered = nunjucks.renderString(cheerio.html(children), {
-              // This is to prevent the nunjuck call from converting {{hostBaseUrl}} to an empty string
-              // and let the hostBaseUrl value be injected later.
-              hostBaseUrl: '{{hostBaseUrl}}',
-              baseUrl: newBaseUrl,
-            }, { path: filePath });
-            element.children = cheerio.parseHTML(rendered, true);
-            cheerio.prototype.options.xmlMode = true;
-          }
-        }
+  /**
+   * @throws Will throw an error if a non-absolute path or path outside the root is given
+   */
+  static calculateNewBaseUrls(filePath, root, lookUp) {
+    if (!path.isAbsolute(filePath)) {
+      throw new Error(`Non-absolute path given to calculateNewBaseUrls: '${filePath}'`);
+    }
+    if (!pathIsInside(filePath, root)) {
+      throw new Error(`Path given '${filePath}' is not in root '${root}'`);
+    }
+    function calculate(file, result) {
+      if (file === root) {
+        return { relative: path.relative(root, root), parent: root };
+      }
+      const parent = path.dirname(file);
+      if (lookUp.has(parent) && result.length === 1) {
+        return { relative: path.relative(parent, result[0]), parent };
+      } else if (lookUp.has(parent)) {
+        return calculate(parent, [parent]);
       }
+      return calculate(parent, result);
     }
-    delete element.attribs[ATTRIB_INCLUDE_PATH];
+
+    return calculate(filePath, []);
+  }
+
+  static calculateBoilerplateFilePath(pathInBoilerplates, asIfAt, config) {
+    const { parent, relative } = Parser.calculateNewBaseUrls(asIfAt, config.rootPath, config.baseUrlMap);
+    return path.resolve(parent, relative, BOILERPLATE_FOLDER_NAME, pathInBoilerplates);
   }
-  delete element.attribs[ATTRIB_CWF];
-  return element;
-};
 
-Parser.prototype._rebaseReferenceForStaticIncludes = function (pageData, element, config) {
-  if (!config) {
-    return pageData;
+  static createErrorNode(element, error) {
+    const errorElement = cheerio.parseHTML(utils.createErrorElement(error), true)[0];
+    return Object.assign(element, _.pick(errorElement, ['name', 'attribs', 'children']));
   }
 
-  if (!pageData.includes('{{baseUrl}}')) {
-    return pageData;
+  static createEmptyNode() {
+    const emptyElement = cheerio.parseHTML('<div></div>', true)[0];
+    return emptyElement;
   }
 
-  const filePath = element.attribs[ATTRIB_INCLUDE_PATH];
-  const fileBase = calculateNewBaseUrls(filePath, config.rootPath, config.baseUrlMap);
-  if (!fileBase.relative) {
-    return pa
```

Comment:
Don't think it should be resolved in this PR, but can this be declared in the constructor?

What do you think @yamgent?

Reply:
It will probably take a bit of work to move it to the constructor, since it is likely that the original author intentionally makes it globally scoped.

Reply:
So yes let's leave it to another PR.